com.unity.netcode.gameobjects/Runtime/Messaging/MessagingSystem.cs

```diff
@@ -226,6 +259,66 @@ private bool CanReceive(ulong clientId, Type messageType, FastBufferReader messa
             return true;
         }
 
+        internal void ReorderMessage(int desiredOrder, uint targetHash)
+        {
+            if (desiredOrder < 0)
+            {
+                // invalid message re-ordering request. Ignore.
+                return;
+            }
+
+            if (desiredOrder < m_ReverseTypeMap.Length &&
+                XXHash.Hash32(m_ReverseTypeMap[desiredOrder].FullName) == targetHash)
+            {
+                // matching positions and hashes. All good.
+                return;
+            }
+
+            // Since the message at `desiredOrder` is not the expected one,
+            // insert an empty placeholder and move the messages down
+            var typesAsList = m_ReverseTypeMap.ToList();
+            typesAsList.Insert(desiredOrder, null);
+            m_ReverseTypeMap = typesAsList.ToArray();
+            var handlersAsList = m_MessageHandlers.ToList();
+            handlersAsList.Insert(desiredOrder, null);
+            m_MessageHandlers = handlersAsList.ToArray();
+
+            // we added a dummy message, bump the end up
+            m_HighMessageType++;
+
+            int position = desiredOrder;
+            bool found = false;
+            while (position < m_ReverseTypeMap.Length)
+            {
+                if (m_ReverseTypeMap[position] != null &&
+                    XXHash.Hash32(m_ReverseTypeMap[position].FullName) == targetHash)
+                {
+                    found = true;
+                    break;
+                }
+
+                position++;
+            }
+
+            if (found)
+            {
+                // Copy the handler and type to the right index
+                m_ReverseTypeMap[desiredOrder] = m_ReverseTypeMap[position];
+                m_MessageHandlers[desiredOrder] = m_MessageHandlers[position];
+
+                // Shift back the remaining messages
+                typesAsList = m_ReverseTypeMap.ToList();
```

Comment:
Why are you doing array->list->array conversions again? Shouldn't you either always have them be lists as in a comment I made above, or else convert array->list, do ALL operations in this method on the lists, then just do 1 conversion back to array?

Reply:
Well, it was either:
1. rework the whole code. I dislike doing that on point releases. Especially when the original code author is away for a while
2. write hard-to-read iterative in-array shifting of the values. Would be faster at runtime, though.
3. write pretty direct, clear semantic, list insert and remove. But yeah, on the performance front, it's lacking. But this happens once at connect time. Even with many messages, it will finish in less than a milli.

I picked 3, mainly so intent was clear. If you insist, I can do 1. or 2. But I really feel, although the code looks to do repeating operations, it will be easier to maintain.

Reply:
You can do #3 with 2 fewer Array allocations and 2 fewer List allocations per message. I'll add some suggested changes.

Reply:
Check out https://gist.github.com/JesseOlmer/92c8188693d2ec833f4ea4987fafe64c... I THINK it should work - it passes your tests, but it also passed your tests when it initially has a bug in it so :shrug: